docs: Add example with hover path and search box

dsmedia opened the pull request:
- point paths on hover
- search box to filter by country name

Relates to [#2980](https://github.com/vega/altair/issues/2980)

TODO:
- Investigate and potentially correct North Korea and South Korea data in vega_datasets' gapminder json, which appears to have swapped labels.

<!-- Click "Preview" to read this message; then delete it. -->

Thanks for contributing to Altair!

Please follow these guidelines when submitting your PR:

- Describe the purpose of the changes in PR, so that it is easy to understand the implication of the suggested changes.
    - Here is a [helpful article about writing effective PR descriptions](https://medium.com/@greenberg/writing-pull-requests-your-coworkers-might-enjoy-reading-9d0307e93da3).
- Include unit tests and documentation for new features
- Ensure that the title is a concise [semantic commit message](https://www.conventionalcommits.org/) (e.g. "feat: Add `embed_options` to charts").
    - Append `!` if the change is breaking (e.g. "fix!: Raise error when `embed_options` is `None`")

mattijn commented:
Thanks for this PR! Looks great, but a little bit big for my screen and adoption in the example page.

Can you add `nearest=True` within the following line?:
```python
hover = alt.selection_point(on='mouseover', fields=['country'], empty=False)
```

Another thing, it is recommended to use the approach as is documented in https://github.com/vega/altair/blob/main/CONTRIBUTING.md#adding-examples, especially:

> Once you have an example you would like to add there are a few guide lines to follow. Every example should:
> - have a `arguments_syntax` and `methods_syntax` implementation. Each implementation must be saved as a stand alone script in the `tests/examples_arguments_syntax` and `tests/examples_methods_syntax` directories.

Thanks again for this PR and hopefully you are able to improve this PR using these these suggestions.

dsmedia commented:
The addition of nearest=True adds a nice visual effect of keeping one path visible at all times once the mouse enters the chart. This serves as a reminder that there is a time element that can be interacted with, not only a cross-section of countries. Without this, a user may not realize there is the option to see both panel and time-series data at the same time, unless they happen to mouse over directly on a mark.

One concern about adding nearest=True is that it somewhat alters the experience of mousing over a country and then trying to scroll through the year slider while keeping in focus that country's path over time, as the movement of the mouse toward the slider tends to trigger *another* country's path to become highlighted as it passes near to another country.

Any thoughts on this? One thought is to keep nearest as False but then to have one country's path shown in the chart's initial state by default. I'm not sure how to implement that, though.

mattijn commented:
Good point. Maybe like this?
```python
hover = alt.selection_point(on='click', fields=['country'], empty=False, value='Egypt')
```
Its not really a `hover` anymore like this though..

joelostblom commented on tests/examples_methods_syntax/scatter_point_paths_hover.py:
```diff
@@ -0,0 +1,95 @@
+x_slider = alt.binding_range(min=1955, max=2005, step=5, name='Year: ')
+x_select = alt.selection_point(name="x_select", fields=['year'], bind=x_slider, value=1980)
+
+# Hover selection
+hover = alt.selection_point(on='mouseover', fields=['country'], empty=False)
```
```suggestion
# Hover selection
hover = alt.selection_point(on='mouseover', fields=['country'], empty=False)
# A separate hover for the points since these need empty=True
hover_point_opacity = alt.selection_point(on='mouseover', fields=['country'])
```

I think having the points fade out on hover makes it easier to see the line. See change below for where this is used.

joelostblom commented on tests/examples_methods_syntax/scatter_point_paths_hover.py:
```diff
@@ -0,0 +1,95 @@
+base = alt.Chart(source).encode(
+    x=alt.X('fertility:Q', scale=alt.Scale(zero=False), title='Fertility'),
+    y=alt.Y('life_expect:Q', scale=alt.Scale(zero=False), title='Life Expectancy'),
+    color=alt.Color('region_name:N').legend(title='Region'),
+    detail='country:N'
+).transform_calculate(
+    region_name="{'0': 'South Asia', '1': 'Europe & Central Asia', '2': 'Sub-Saharan Africa', '3': 'America', '4': 'East Asia & Pacific', '5': 'Middle East & North Africa'}[datum.cluster]"
```
```suggestion
    color=alt.Color('region:N', title='Region'),
    detail='country:N'
).transform_calculate(
    region="""{
        '0': 'South Asia',
        '1': 'Europe & Central Asia',
        '2': 'Sub-Saharan Africa',
        '3': 'America',
        '4': 'East Asia & Pacific',
        '5': 'Middle East & North Africa'
    }[datum.cluster]"""

```

Setting title the same way as for x and y

joelostblom commented on tests/examples_methods_syntax/scatter_point_paths_hover.py:
```diff
@@ -0,0 +1,95 @@
+    baseline='middle',
+    fontSize=96,
+    opacity=0.2,
+    x='width',
+    y='height',
+    dx=-400,
+    dy=-300
```
```suggestion
    x=alt.expr('width/2'),
    y=alt.expr('height/2'),
```

joelostblom reviewed (changes requested):
This is great, thank you @dsmedia ! I made a few inline comments with suggestions. Regarding `nearest` I think the current behavior where the country line is not highlighted when the slider is moved is totally fine. I see that as a drill down action to see the entire trajectory of a single (or a few) countries, whereas the slider is more for seeing how all countries evolve over time as a group.

dsmedia commented:
Thanks for all the feedback @mattijn and @joelostblom. I've incorporated as best I could, while also fixing an issue with hover lines connecting out of sequential order, switched mark_line to mark_trail (which I think gives an intuitive sense of the sequential order) and made some other stylistic modifications, such as setting the country label on hover to appear above the point with the largest y-value for the country. I wasn't sure if there was a consensus about adding nearest=True but chose to leave it out as it seemed to make the hovering experience a bit too bouncy. But I'm happy to add it back in if that's the consensus. I will add the other syntax once the iteration on this vesion is complete. 

@joelostblom could you check the behavior when the enter key is hit on the search box? I'm not sure what's expected but it seems to clear the visualization when I tested it in an interactive window in vscode. Also on the search box, do there happen to be any configuration options (autocomplete?) for the search box that are worth demonstrating here, since I think it'll be its first appearance in an example?

joelostblom commented on tests/examples_methods_syntax/scatter_point_paths_hover.py:
```diff
@@ -0,0 +1,149 @@
+    groupby=['country']
+)
+
+background_year = alt.Chart(source).mark_text(
+    align='center',
+    baseline='middle',
+    fontSize=96,
+    opacity=0.2
+).encode(
+    text='year:O',
+    x=alt.X('width:Q', axis=None, scale=None),
+    y=alt.Y('height:Q', axis=None, scale=None)
+).transform_filter(
+    x_select
+).transform_aggregate(
+    year='max(year)'
+).transform_calculate(
+    width='width/2',
+    height='height/2'
+)
```
```suggestion
background_year = alt.Chart(source).mark_text(
    align='center',
    baseline='middle',
    fontSize=96,
    opacity=0.2
).encode(
    text='year:O',
).transform_filter(
    x_select
).transform_aggregate(
    year='max(year)'
)
```

The year text appear centered for me by default, is there another reason this is needed?

joelostblom commented on tests/examples_methods_syntax/scatter_point_paths_hover.py:
```diff
@@ -0,0 +1,149 @@
+
+# Year labels
+year_labels = base.mark_text(align='left', dx=5, dy=-5, fontSize=14).encode(
+    text='year:O', color=alt.value('#222222')
```
```suggestion
    text='year:O',
    color=alt.value('#222222')
```

joelostblom commented on tests/examples_methods_syntax/scatter_point_paths_hover.py:
```diff
@@ -0,0 +1,149 @@
+
+# Points that are always visible (filtered by slider and search)
+visible_points = base.mark_circle(size=100).encode(
+    opacity=alt.condition(
+            hover_point_opacity & 
+        alt.expr.test(alt.expr.regexp(search_box, 'i'), alt.datum.country),
+        alt.value(0.8),
+        alt.value(0.1))
```
```suggestion
    opacity=alt.condition(
        hover_point_opacity
        & alt.expr.test(alt.expr.regexp(search_box, 'i'), alt.datum.country),
        alt.value(0.8),
        alt.value(0.1)
    )
```

joelostblom commented on tests/examples_methods_syntax/scatter_point_paths_hover.py:
```diff
@@ -0,0 +1,149 @@
+base = alt.Chart(source).encode(
+    x=alt.X('fertility:Q', scale=alt.Scale(zero=False, nice=False), title='Fertility'),
+    y=alt.Y('life_expect:Q', scale=alt.Scale(zero=False, nice=False), title='Life Expectancy'),
+    color=alt.Color('region:N', title='Region').legend(orient='bottom-left', titleFontSize=14, labelFontSize=12).scale(scheme='dark2'),
```
Use `legend=` and `scale=` here to be consistent with the argument style used in the rest of the file

joelostblom commented on tests/examples_methods_syntax/scatter_point_paths_hover.py:
```diff
@@ -0,0 +1,149 @@
+
+# Base chart
+base = alt.Chart(source).encode(
+    x=alt.X('fertility:Q', scale=alt.Scale(zero=False, nice=False), title='Fertility'),
+    y=alt.Y('life_expect:Q', scale=alt.Scale(zero=False, nice=False), title='Life Expectancy'),
```
I think the default behavior looks better than `nice=False` here. It make the dots less crammed in the top left corner:

nice=False
![image](https://github.com/vega/altair/assets/4560057/baa0ff2f-970b-438c-b2be-8a9494c9c14d)


Default:
![image](https://github.com/vega/altair/assets/4560057/5db9a28a-3827-40da-a0fd-10dd18f83daa)

joelostblom commented on tests/examples_methods_syntax/scatter_point_paths_hover.py:
```diff
@@ -0,0 +1,149 @@
+).properties(
+    width=500,
+    height=500,
+    padding=35  # Padding ensures labels fit
```
```suggestion
    padding=10  # Padding ensures labels fit
```

If we remove `nice=False` as per my comment above, I think we are fine with smaller padding here:

![image](https://github.com/vega/altair/assets/4560057/fff67278-e006-4b69-9b29-b0410f979e63)

joelostblom reviewed (changes requested):
Great idea with the trail to indicate direction! I only have a few minor comments, so feel free to address those as you see fit and add the methods syntax version of the example also and then I think we can merge.

Regarding the behavior of the search box when enter is hit, I see the same as you both in VS Code and JupyterLab. I'm not sure how to change it, but we don't need to worry about it here, should probably be fixed on the Vega side of things.

Regarding autocomplete, Vega would need to add support for the `datalist` parameter and a way to pass a datalist to search input. I had opened an issue for it in the past here https://github.com/vega/vega/issues/3702, but didn't get any insights

dsmedia commented:
I see above "Some checks were not successful" with "lint / ruff-mypy (pull_request)" failing. Is there something I should do to fix that?

dangotbanned commented:
> I see above "Some checks were not successful" with "lint / ruff-mypy (pull_request)" failing. Is there something I should do to fix that?

Will be fixed once https://github.com/vega/altair/pull/3463 is merged

joelostblom commented on tests/examples_arguments_syntax/scatter_point_paths_hover.py:
```diff
@@ -0,0 +1,148 @@
+    color=alt.value('black'),
+    opacity=alt.condition(hover, alt.value(1), alt.value(0))
+).transform_window(
+    rank='rank(life_expect)', # 
```
```suggestion
    rank='rank(life_expect)',
```

joelostblom commented on tests/examples_methods_syntax/scatter_point_paths_hover.py:
```diff
@@ -0,0 +1,141 @@
+    color=alt.value('black'),
+    opacity=alt.condition(hover, alt.value(1), alt.value(0))
+).transform_window(
+    rank='rank(life_expect)', # 
```
```suggestion
    rank='rank(life_expect)',
```

joelostblom reviewed (approved):
This is looking great! Thanks for all your work in putting this example together and unveiling the data error as well @dsmedia ! Tests should pass now so I will go ahead and merge when they finish running!
